File: python/src/ux_channel/arch/stamps.py
"""Host-owned stamp table for invoke refs. Peer never owns authority."""

from __future__ import annotations

import secrets
import threading
from dataclasses import dataclass
from typing import Dict, Optional, Set


@dataclass
class Stamp:
    stamp_id: str
    kind: str
    methods: Set[str]
    session_id: str
    gen: int


class StampTable:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._by_session: Dict[str, Dict[str, Stamp]] = {}

    def grant(
        self,
        session_id: str,
        gen: int,
        kind: str,
        methods: Set[str],
        stamp_id: Optional[str] = None,
    ) -> Stamp:
        sid = stamp_id or secrets.token_urlsafe(10)
        st = Stamp(
            stamp_id=sid,
            kind=kind,
            methods=set(methods),
            session_id=session_id,
            gen=gen,
        )
        with self._lock:
            self._by_session.setdefault(session_id, {})[sid] = st
        return st

    def get(self, session_id: str, stamp_id: str, gen: int) -> Optional[Stamp]:
        with self._lock:
            st = self._by_session.get(session_id, {}).get(stamp_id)
            if st is None or st.gen != gen:
                return None
            return st

    def allows(self, session_id: str, stamp_id: str, gen: int, method: str) -> bool:
        st = self.get(session_id, stamp_id, gen)
        if st is None:
            return False
        return method in st.methods or "*" in st.methods

    def clear_session(self, session_id: str) -> None:
        with self._lock:
            self._by_session.pop(session_id, None)

    def on_revoke(self, session_id: str) -> None:
        self.clear_session(session_id)

File: rust/src/stamps.rs
//! Host-owned stamp table for invoke refs. Peer never owns authority.

use hmac::{Hmac, Mac};
use sha2::Sha256;
use std::collections::{HashMap, HashSet};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

type HmacSha256 = Hmac<Sha256>;

#[derive(Debug, Clone)]
pub struct Stamp {
    pub stamp_id: String,
    pub kind: String,
    pub methods: HashSet<String>,
    pub session_id: String,
    pub gen: u64,
}

pub struct StampTable {
    inner: Mutex<HashMap<String, HashMap<String, Stamp>>>,
}

impl Default for StampTable {
    fn default() -> Self {
        Self {
            inner: Mutex::new(HashMap::new()),
        }
    }
}

impl StampTable {
    pub fn grant(
        &self,
        session_id: &str,
        gen: u64,
        kind: &str,
        methods: impl IntoIterator<Item = impl Into<String>>,
    ) -> Stamp {
        let stamp_id = fresh_id(session_id, gen, kind);
        let st = Stamp {
            stamp_id: stamp_id.clone(),
            kind: kind.into(),
            methods: methods.into_iter().map(Into::into).collect(),
            session_id: session_id.into(),
            gen,
        };
        let mut g = self.inner.lock().expect("stamps");
        g.entry(session_id.into())
            .or_default()
            .insert(stamp_id, st.clone());
        st
    }

    pub fn allows(&self, session_id: &str, stamp_id: &str, gen: u64, method: &str) -> bool {
        let g = self.inner.lock().expect("stamps");
        match g.get(session_id).and_then(|m| m.get(stamp_id)) {
            Some(st) if st.gen == gen => st.methods.contains(method) || st.methods.contains("*"),
            _ => false,
        }
    }

    pub fn on_revoke(&self, session_id: &str) {
        self.inner.lock().expect("stamps").remove(session_id);
    }
}

fn fresh_id(session_id: &str, gen: u64, kind: &str) -> String {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let mut mac = HmacSha256::new_from_slice(b"uxc-stamp").expect("hmac");
    mac.update(session_id.as_bytes());
    mac.update(&gen.to_be_bytes());
    mac.update(kind.as_bytes());
    mac.update(&nanos.to_be_bytes());
    let out = mac.finalize().into_bytes();
    hex::encode(&out[..8])
}
